helix-term/src/commands.rs

```diff
@@ -3594,6 +3594,39 @@ fn normal_mode(cx: &mut Context) {
 
     doc.mode = Mode::Normal;
 
+    let doc_changes = doc.changes().changes();
+    let text = doc.text().slice(..);
+    let pos = doc.selection(view.id).primary().cursor(text);
+    let mut can_restore_indent = false;
+
+    // Will remove whitespaces when user opens a new line(may additional insert only tabs or spaces)
+    // and goto normal mode.
+    //
+    // NOTE: when open a new line, changes will only contains 3 elements:
+    // retain(pos) -> insert(something) -> retain(remaining_length), the last retain is useless, we
+    // only concerned about the first two changes.
+    if doc_changes.len() == 3 {
+        if let helix_core::Operation::Retain(move_pos) = doc_changes[0] {
+            if let helix_core::Operation::Insert(ref inserted_str) = doc_changes[1] {
```

Comment:
Why not `if let [Retain(move_pos), Insert(inserted_str), _] = doc_changes {` directly?

Reply:
Done.  It's simply because I don't remember the pattern..